Add user profile heatmap visualization for contributions

kcne opened the pull request:
### Description
This PR addresses #5373 by adding a heatmap visualization on user profile pages to display daily contribution activity over the past year. 

**Key changes include:**
- Added a heatmap container in the user profile view (`users/show.html.erb`).
- Implemented JavaScript logic (`heatmap.js`) using CalHeatmap to render the heatmap dynamically based on contribution data.
- Updated localization (`config/locales/en.yml`) to include tooltips and month/weekday labels for the heatmap.
- Added tests in `users_controller_test.rb` to validate heatmap data handling.

Screenshots:
<img width="951" alt="Screenshot 2024-12-16 at 16 35 58" src="https://github.com/user-attachments/assets/878068ee-c0cd-450d-b3e8-010a9c75575f" />
<img width="964" alt="Screenshot 2024-12-16 at 16 37 14" src="https://github.com/user-attachments/assets/2ae8ff02-0416-41fe-aeeb-d91cc6584a67" />
<img width="426" alt="Screenshot 2024-12-16 at 16 36 25" src="https://github.com/user-attachments/assets/ca4d4b80-cdbf-4b20-a5e6-602ef3ec07ce" />

### How has this been tested?
- Verified heatmap rendering with mock data for different users and activity levels.
- Tested responsiveness and theme support (light/dark mode).
- Checked correct localization of tooltips, month names, and weekday labels.
- Ran automated tests to ensure proper heatmap data handling in the controller.

Let me know if you need further tweaks or enhancements!

HolgerJeromin commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,116 @@
+  } catch (error) {
+    console.error("Error during Cal-Heatmap initialization:", error);
+  }
+});
```
```suggestion
}, {once: true});
```

kcne replied:
Fixed. Thanks for the suggestion

HolgerJeromin commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,116 @@
+    const cal = new CalHeatmap();
+    const applicationData = $("head").data();
+    const locale = applicationData.locale;
+    const theme = getTheme();
```
We can watch for color theme with 
https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/change_event

kcne replied:
Refactored the code accordingly so it takes into account if the theme of the browser is changed. Thank you for pointing this out

HolgerJeromin commented:
Oh, I forgot about this. 
But the user can now set his "Preferred Website Color Scheme" which should be the basis of the theme not the browser setting.

gravitystorm commented:
What's the source(s) of all the javascript in `vendor/`? I would expect to see `Vendorfile` updated, or preferably if these are sourced from e.g. node modules, then they should be included in package.json and added using the assets pipeline. If any of these are source files, i.e. written just for this project, then they should be in app/.

Also, how does the inclusion of popper interact with the popper already included? We end up with `//= require popper` twice in application.js.

gravitystorm commented:
> What's the source(s) of all the javascript in `vendor/`?

OK, I found the answer for the cal-heatmap part of the code, which is available as an npm module. In which case, using package.json will allow us to upgrade to new versions automatically, and would be the preferred option.

tordans commented:
UX thoughts:
- Do we want users to be able to disable this feature in their profile? – I am not a fan of this kind of configurations, but we do give a lot more visibility to once activity with this that we had before and the page is public (AKA no login required).
- How do we want to handle the "never contributed" case? Showing the heatmap in this case would suggest "this user might have contributed changes before that time…". – I think it would be best to not show it or show a placeholder instead.

Code comments:
- Do we want to add those JS files to the application js or create a separate contribution_heatmap js for this? https://github.com/openstreetmap/openstreetmap-website/pull/5402/files#diff-0e145ba000e96b1a45d230275e350e4e8514efb4db7dbcfa228fe1003896c9b4R18-R22

tomhughes commented:
I've tested the performance of the underlying query and using the user with the most changesets in the last year (just over 50 thousand) it took a fraction over one second which is probably acceptable.

That result (which was a full 365 days of data) serialised to just under 10Kb which is a fair chunk to be caching but hopefully most entries will be smaller than that.

kcne commented:
> Oh, I forgot about this. But the user can now set his "Preferred Website Color Scheme" which should be the basis of the theme not the browser setting.

Thank you for the suggestion. I’ll refactor accordingly to ensure alignment with the preferred color scheme setting.



> It is a very minor nitpick, but is it possible to dynamically determine what's the first weekday based on the user's locale and adjust the heatmap accordingly? [Intl.Locale.prototype.getWeekInfo()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/getWeekInfo) could be used. It isn't currently available in Firefox but all other browsers have it.
> 
> P.S. Added [bug 1937867](https://bugzilla.mozilla.org/show_bug.cgi?id=1937867) to Firefox's bug tracker.

I tried dynamically adjusting the start and end weekdays using [ghDay](https://cal-heatmap.com/docs/options/subDomain#ghday) as the subdomain, but it rounds to the first and last weeks of the month, making implementation tricky. Accommodating this would require significant effort, which may not be worth it for such a minor adjustment.



> What's the source(s) of all the javascript in `vendor/`? I would expect to see `Vendorfile` updated, or preferably if these are sourced from e.g. node modules, then they should be included in package.json and added using the assets pipeline. If any of these are source files, i.e. written just for this project, then they should be in app/.

To be honest I wasn’t  too familiar with how our assets pipeline works,so I just added the `cal-heatmap` library and its dependencies in `vendor/` and added required code in `manifest.js` and `application.js` to make it load correctly. If using `package.json` is preferred, I can update the PR accordingly. I'm not sure if there is some additional config needed for this to be accomplished or would updating the `package.json` file would be enough.

Thank you for pointing out the existing `popper` inclusion. I’ll remove the duplicate entry from `application.js` and the `vendor` folder.

AntonKhorev commented:
> Separate JS Files: There's already a heatmap.js file containing heatmap-related functionality, included in application.js for precompilation.

How is it included in `application.js` if you're including `heatmap.js` [here](https://github.com/openstreetmap/openstreetmap-website/pull/5402/files#diff-78ff736409d758722403cce31873ba803b3fb526d37398ca21caa557b54dfd95R3)? But `d3` and `cal-heatmap` are added to `application.js`, bringing its size up to ~400kb from ~250kb.

AntonKhorev commented:
CSP violations, days of week won't be visible if not resolved:

> Content-Security-Policy: (Report-Only policy) The page’s settings would block an inline style (style-src-attr) from being applied

![image](https://github.com/user-attachments/assets/63e176b4-22d8-4238-9d3c-101a1cd06332)

AntonKhorev commented:
> I've updated the code to account for the "Preferred Website Color Scheme." Theme settings should now be working correctly.

It's not going to work correctly because our *Auto* option follows `prefers-color-scheme` which can change any time. If it does, you'll get for example this:

![image](https://github.com/user-attachments/assets/77ad7234-a4e0-4441-a321-6f723f5a826a)

Looks like cal-heatmap doesn't have a proper auto scheme ["as not all websites support dark/light mode"](https://cal-heatmap.com/docs/options/theme).

You'd have to then setup a listener like [this](https://github.com/openstreetmap/openstreetmap-website/pull/4777/commits/e35141b2d1c1733d9a15a01c08b796e4edda6b4e#diff-4fe14e869cf17b068e1bea009a55b094fed13f0cad85337cf4769f32b40a01b5R11-R13) but without using Leaflet of course and redraw the heatmap.

Same is true for #5426 by the way.

hlfan commented:
> You'd have to then setup a listener like [this](https://github.com/openstreetmap/openstreetmap-website/pull/4777/commits/e35141b2d1c1733d9a15a01c08b796e4edda6b4e#diff-4fe14e869cf17b068e1bea009a55b094fed13f0cad85337cf4769f32b40a01b5R11-R13) but without using Leaflet of course and redraw the heatmap.
> 
> Same is true for #5426 by the way.

Any other change to the theme of the page without reloading will break most of the CSS theming anyway since the stylesheets were split in #5362 by the way.

It would be great to have a global leaflet-independent watcher where clients can add themselves to a callback list.

AntonKhorev commented:
> Any other change to the theme of the page without reloading will break most of the CSS theming anyway since the stylesheets were split in #5362 by the way.

You don't need to change the theme of the page, neither here, nor in https://github.com/openstreetmap/openstreetmap-website/pull/5426.

> It would be great to have a global leaflet-independent watcher where clients can add themselves to a callback list.

Leaflet code is just a wrapper for `addEventListener`. You can do the same with jQuery.

hlfan commented:
> You don't need to change the theme of the page

I did that for testing, but since listening increased the complexity unnecessarily, I left that for a second pass.

But suppose the site needs a theme watcher anyway because the cal-heatmap doesn't want to let CSS do the color mixing theme-agnostically. In that case, instead of having two listeners, a common solution that includes MutationObservers outside of jQuery would be the one with the least overhead.

AntonKhorev commented:
Why MutationObservers?

AntonKhorev commented:
Also the conditions under which watching `prefers-color-scheme` are different for this PR and https://github.com/openstreetmap/openstreetmap-website/pull/5426. Here you need to watch if the website theme is auto, in https://github.com/openstreetmap/openstreetmap-website/pull/5426 you need to watch when the website and the map themes are both auto.

hlfan commented:
The MutationObserver thought came from my tinkering with the site locally wanting to reduce the reloads, forgetting that a toggle if added would already have onchange events.

> So maybe you don't want one global watcher.

Yeah, I kinda missed that... Still, having a similar style on such similar functions and having them moved to some unified location would be a nice thing to have, currently, they could be much more different. But that's more of a refactor thing if both are merged.

kcne commented:
> > Separate JS Files: There's already a heatmap.js file containing heatmap-related functionality, included in application.js for precompilation.
> 
> How is it included in `application.js` if you're including `heatmap.js` [here](https://github.com/openstreetmap/openstreetmap-website/pull/5402/files#diff-78ff736409d758722403cce31873ba803b3fb526d37398ca21caa557b54dfd95R3)? But `d3` and `cal-heatmap` are added to `application.js`, bringing its size up to ~400kb from ~250kb.

I've moved the import to `heatmap.js`. Should be resolved now.



> > I've updated the code to account for the "Preferred Website Color Scheme." Theme settings should now be working correctly.
> 
> It's not going to work correctly because our _Auto_ option follows `prefers-color-scheme` which can change any time. If it does, you'll get for example this:
> 
> ![image](https://private-user-images.githubusercontent.com/4158490/399146698-77ad7234-a4e0-4441-a321-6f723f5a826a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY4NDQ5NzMsIm5iZiI6MTczNjg0NDY3MywicGF0aCI6Ii80MTU4NDkwLzM5OTE0NjY5OC03N2FkNzIzNC1hNGUwLTQ0NDEtYTMyMS02ZjcyM2Y1YTgyNmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDExNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMTRUMDg1MTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjRiNmEwNGRhZWUwYTllMDZjMDRlNGM5Yjg5OTViMjQwNzM1OGE0NDg2YWIzYWZkOTMzNTliYTBmNTU5NGE3YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.09evvZz2WBft87-EVJs6CR7YWBIyJxyH79iJhJvn2ho)
> 
> Looks like cal-heatmap doesn't have a proper auto scheme ["as not all websites support dark/light mode"](https://cal-heatmap.com/docs/options/theme).
> 
> You'd have to then setup a listener like [this](https://github.com/openstreetmap/openstreetmap-website/pull/4777/commits/e35141b2d1c1733d9a15a01c08b796e4edda6b4e#diff-4fe14e869cf17b068e1bea009a55b094fed13f0cad85337cf4769f32b40a01b5R11-R13) but without using Leaflet of course and redraw the heatmap.
> 
> Same is true for #5426 by the way.

Added the logic to handle the auto prefered-color-changes. Should be resolved now as well.





> CSP violations, days of week won't be visible if not resolved:
> 
> > Content-Security-Policy: (Report-Only policy) The page’s settings would block an inline style (style-src-attr) from being applied
> 
> ![image](https://private-user-images.githubusercontent.com/4158490/399146508-63e176b4-22d8-4238-9d3c-101a1cd06332.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY4NDQ5NzMsIm5iZiI6MTczNjg0NDY3MywicGF0aCI6Ii80MTU4NDkwLzM5OTE0NjUwOC02M2UxNzZiNC0yMmQ4LTQyMzgtOWQzYy0xMDFhMWNkMDYzMzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDExNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMTRUMDg1MTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTBmMzk1ZDkzMGZjMDJlMWEwMzliZjg4OTEyYWY1NGQ2MmZjNDZiN2Y0MjRhYmM5YTFkNTU5ZjI3MTdhZDY2MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.NOccBT9pka3XwEZsiLGH3Z_6Oard3C0UcPaL8sTn8wk)

I'm working on resolving this CSP violations. They are from `CalendarLabel` plugin. Right now I'm trying to generate nonce for the .js file and use that to validate and transpass the CSP violation. I'm not sure if it is the right approach for this since I never had experience dealing with similar stuff before. If there is a more elegant and compact solution, please feel free to point it out.

Thank you.

hlfan commented:
> Added the logic to handle the auto prefered-color-changes. Should be resolved now as well.

For the wrong one though. Another copy of the html's bs-theme attribute isn't necessarily helpful, and the media query is still not being listened to.

kcne commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,125 @@
+  ]);
+}, { once: true });
+
+function getThemeFromColorScheme(colorScheme) {
+  if (colorScheme === "auto") {
+    return window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light";
+  }
+  return colorScheme; // Return "light" or "dark" directly if specified
+}
```
I appreciate your feedback @hlfan. From my understanding, the primary issue was that when the “Auto” option is enabled, changes in the system’s prefers-color-scheme weren’t being reflected. With the new matchMedia logic, however, we’re directly listening for those changes, which triggers a re-draw of the heatmap whenever the user’s color scheme switches.

If you’re still seeing a scenario where the theme isn’t updating correctly, please let me know so I can investigate further. Otherwise, the current solution should correctly handle automatic theme changes.

HolgerJeromin replied:
`getThemeFromColorScheme` returns the correct theme. Yes. 
But it is only called on load (`DOMContentLoaded`) and not when the user changes operating system theme.

Missing something like:
`window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", xy);`

hlfan commented:
@HolgerJeromin I'd recommend `window.matchMedia("...").onchange = xy` due to the [depreciation](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/addListener) of `matchMedia.addEventListener`.

kcne commented:
I need to test this once again and get back to you. I guess you are right since there is some strange behaviour right now that I'm currently looking into.

HolgerJeromin commented:
@hlfan  This deprecates `addListener` (old IE invention) not `addEventListener`

kcne commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,134 @@
+  let cal = new CalHeatmap();
+  renderHeatmap(theme);
+
+  if (colorScheme === "auto") {
+    mediaQuery.addEventListener("change", (e) => {
+      const newTheme = e.matches ? "dark" : "light";
+      renderHeatmap(newTheme);
+    });
+  }
```
@hlfan Sorry about confusion here. 
In the process of rebasing the commit for latest changes I forgot to pop stashed changes so the file version didn't reflect last changes. Thanks for catching that and pointing it out. 

This is the updated version. It's working as expected in my firefox browser locally now. Please let me know if there is something else I should modify here. 

Thank you.

simonpoole commented:
Nice ....... but, sorry to be a spoil sport here, this changes the information available on the user page significantly, the likely solution would be to make it available only to logged in users. 

Both the LWG and EWG should be informed and https://wiki.openstreetmap.org/wiki/GDPR/Affected_Services#User_page_(done) updated to reflect the additional feature.

PS: I'm not quite sure why the user page item is marked as done, when it obviously isn't, but that is OT.

AntonKhorev commented:
@simonpoole This is not something you should worry about because it will take one `if` statement to remove the entire thing unlike let's say feeds.

AntonKhorev commented on app/assets/javascripts/application.js:
```diff
@@ -15,6 +15,11 @@
+//= require cal-heatmap
+//= require calendar-label
+//= require popper
+//= require tooltip
```
You don't need to put these scripts here because you remove them later.

kcne replied:
This should be removed. Thanks.

AntonKhorev replied:
This should be removed but this is not yet removed because this commit is still in the branch.

kcne replied:
You were right, I was overriding changes in the later commit, instead of just updating this commit. It's fixed now. Thank you.

AntonKhorev commented on app/controllers/users_controller.rb:
```diff
@@ -53,6 +53,23 @@ def show
        (@user.visible? || current_user&.administrator?)
       @title = @user.display_name
+
+      one_year_ago = 1.year.ago
```
Is this done to freeze the starting point because we're not sure when the block below runs? Then why is the same not done with the ending point? And shouldn't it be truncated to a day?

kcne replied:
This wasn’t about freezing; it should run each time the user profile page is accessed. I guess I left it like that by accident. The idea was to fetch the changeset data from one year ago up until the current date. The thought was that starting one year ago would naturally include changesets up to this moment, effectively capturing the data for the past year.

However, I’ve now refactored the code to explicitly set both the starting and ending points. This makes the date range more declarative and ensures better clarity and consistency, while also eliminating potential ambiguities.

kcne replied:
Just double check if it looks alright before resolving comment. Thank you.

AntonKhorev commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,134 @@
+      scale: {
+        color: {
+          type: "threshold",
+          range: currentTheme === "dark" ? rangeColors : rangeColors.reverse(),
```
`rangeColors.reverse()` reverses the array in place, probably not what you want.

HolgerJeromin replied:
```suggestion
          range: currentTheme === "dark" ? rangeColors : Array.from(rangeColors).reverse(),
```
would solve that.

kcne replied:
Resolved now. Thanks.

AntonKhorev commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,134 @@
+}
+
+function getTooltipText(date, value, locale) {
+  const jsDate = new Date(date);
+  const translations = I18n.translations[locale] || I18n.translations.en;
+  const dateTranslations = translations && translations.date;
+  const monthNames = dateTranslations && dateTranslations.month_names;
+
+  const months = monthNames || [];
+  const monthName = months[jsDate.getMonth() + 1] || `${jsDate.getMonth + 1}.`;
+  const day = jsDate.getDate();
+  const year = jsDate.getFullYear();
+
+  const localizedDate = `${day}. ${monthName} ${year}.`;
```
```suggestion
  const localizedDate = I18n.l("date.formats.long", date);
```

kcne replied:
This simplifies code my code a lot. Thank you very much sir

AntonKhorev commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,134 @@
+
+  const heatmapData = heatmapElement.dataset.heatmap ? JSON.parse(heatmapElement.dataset.heatmap) : [];
+  const colorScheme = heatmapElement.dataset.siteColorScheme || "auto";
+  const locale = $("head").data().locale;
```
If `application.js` is initialized, the locale is available in `I18n.locale`.

AntonKhorev replied:
I think the entire heatmap can be delayed by putting it in turbo frame or maybe some collapsible section (maybe for performance reasons). You need `I18n` initialized before anyway, I don't know how you're going to translate dates correctly otherwise. Although popup dates seem to work even if `I18n` is not initialized before the heatmap because they are constructed later.

kcne replied:
I noticed you opened #5538. Are the changes there related to the initialization concerns mentioned here?  

Regarding delaying the heatmap or using a turbo frame/collapsible section for performance, I’m not sure about the best metrics to determine what's acceptable in terms of load time/size or resource usage. Here's the network tab from Chrome when testing locally with the cache disabled: 

![Screenshot 2025-01-24 at 17 20 16](https://github.com/user-attachments/assets/df1b8c92-ec1c-4ac8-a88e-6a17c2050a5d)  

If there are specific metrics or scenarios you'd recommend testing to ensure this is performant and I18n is properly initialized, I’d be happy to try those out.

Thank you.

simonpoole commented:
> @simonpoole This is not something you should worry about because it will take one `if` statement to remove the entire thing unlike let's say feeds.

Oh, I'm not concerned at all. I was just pointing this out for good form, aka I saw something that I know is wrong, and I'm giving the people here a fighting chance to do the right thing.

kcne commented:
Updated the code again, resolved some issues raised here. Few remarks:

> CSP violations, days of week won't be visible if not resolved:
> 
> > Content-Security-Policy: (Report-Only policy) The page’s settings would block an inline style (style-src-attr) from being applied
> 
> ![image](https://private-user-images.githubusercontent.com/4158490/399146508-63e176b4-22d8-4238-9d3c-101a1cd06332.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc3MzM1NTAsIm5iZiI6MTczNzczMzI1MCwicGF0aCI6Ii80MTU4NDkwLzM5OTE0NjUwOC02M2UxNzZiNC0yMmQ4LTQyMzgtOWQzYy0xMDFhMWNkMDYzMzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjRUMTU0MDUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDhmOTIwM2Y2NzllNWI0ZjIwMmJkYjFjYjY1YWE4NDQzOTJhNGRkM2NhNjhhZDJmMWY5YWZhN2Y4MjQ5ZmQyZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EoL-EszNCUum1gnFkRiUnrIKw93dYJSdCIw2RcnULms)

The issue was caused by the `CalendarLabel` plugin. Given the simplicity of our use case, I opted to convert the labels to use plain HTML/CSS instead of introducing custom CSP policies for controller/view/JS files. Labels should now display correctly, and the CSP violations should be resolved. Please let me know if there’s anything I might have missed.

> Is it the intended effect that nothing gets rendered in case the user made no edits in the span of one year?

The heatmap now renders regardless of whether the user has made edits in the past year. If there’s interest in hiding the heatmap for users with no contributions, I can refactor it to include a fallback or placeholder message instead. Let me know your preference.

tomhughes commented:
> I'm working on resolving this CSP violations. Right now I'm trying to generate nonce for the .js file and use that to validate and transpass the CSP violation. I'm not sure if it is the right approach for this since I never had experience dealing with similar stuff before. If there is a more elegant and compact solution, please feel free to point it out.

Using nonces should be an absolute last resort though it is possible if we need to - it is enabled for styles and the nonce is available in the `csp-nonce` meta tag in the document head.

Much better to use a proper stylesheet and just add and remove classes if you can though.

kcne commented:
> > I'm working on resolving this CSP violations. Right now I'm trying to generate nonce for the .js file and use that to validate and transpass the CSP violation. I'm not sure if it is the right approach for this since I never had experience dealing with similar stuff before. If there is a more elegant and compact solution, please feel free to point it out.
> 
> Using nonces should be an absolute last resort though it is possible if we need to - it is enabled for styles and the nonce is available in the `csp-nonce` meta tag in the document head.
> 
> Much better to use a proper stylesheet and just add and remove classes if you can though.

I agree, Tom. Since it was an external dependency, I removed it and implemented the labels using plain HTML and CSS instead.

AntonKhorev commented on app/assets/config/manifest.js:
```diff
@@ -7,13 +7,17 @@
+//= link_directory ../../../vendor/assets/d3 .js
+//= link_directory ../../../vendor/assets/cal-heatmap .js
+//= link_directory ../../../vendor/assets/cal-heatmap .css
 
```
To address https://github.com/openstreetmap/openstreetmap-website/pull/5402#issuecomment-2549612467 you'd do something like https://github.com/AntonKhorev/openstreetmap-website/commit/7ac349e61d57bf568b55c13a44f0fd0b16261ae3. Except of course you'd start with installing it using `yarn add cal-heatmap` and you wouldn't copy dependencies to `vendor` manually.

kcne replied:
Alright, fixed now and rebased the commits accordingly. Thank you for putting effort into demonstrating how to properly do this!

AntonKhorev commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,104 @@
+
+function getThemeFromColorScheme(colorScheme) {
+  if (colorScheme === "auto") {
+    return window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light";
```
You already have `mediaQuery` defined. It can be used here:
```suggestion
    return mediaQuery.matches ? "dark" : "light";
```

For this to work, declarations have to be moved around. Maybe this function doesn't need to be declared. You already have `if (colorScheme === "auto") ...` above that is not inside a function.

kcne replied:
You were completely right. I tried removing the method, but instead of doing two things in one method or in if statement i just refactored the method and moved it around a bit. Thank you.

AntonKhorev commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,104 @@
+      }
+    }, [
+      [Tooltip, {
+        text: (date, value) => getTooltipText(date, value, locale)
```
`getTooltipText` doesn't have the `locale` parameter. And you probably don't need to pass `locale` around to other places, given that you can read it from `I18n.locale`.

kcne replied:
Fixed. Removed locale altogether since as you pointed out I can use I18n directly. Thank you.

AntonKhorev commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,104 @@
+}
+
+function getMonthNameFromTranslations(locale, monthIndex) {
+  const translations = I18n.translations[locale] || I18n.translations.en;
+  const date = translations && translations.date;
+  const abbrMonthNames = date && date.abbr_month_names;
+
+  const months = abbrMonthNames || [];
+  return months[monthIndex + 1] || "";
```
```suggestion
  return I18n.strftime(new Date(2000, monthIndex, 15), "%b");
```

kcne replied:
This is great, makes code much cleaner and shorter. Removed method altogether and just used the code you provided in the text callback. Thank you.

AntonKhorev commented on app/assets/stylesheets/common.scss:
```diff
@@ -1003,3 +1003,23 @@ img.trace_image {
   }
 }
+
+.heatmap-wrapper {
+  position: relative;
+  height: 130px;
+}
+
+.heatmap-labels {
+position: absolute;
+  top: 2px;
+  left: 17px;
+  padding-top: 21px;
+  z-index: 1;
+  overflow-y: hidden;
+}
+
+.heatmap {
+  margin-left: 50px;
+  height: 100%;
+  z-index: 0;
+}
```
Why is it here and not in `heatmap.css`?

kcne replied:
I decided to remove `heatmap.css` completely since we opted to using npm package and linking stylesheet inside `show.html.erb` directly. Since it's only 18 lines of code I think we can leave it in `common.scss` for now. If you still think I should move it to separate file instead, just let me know.

AntonKhorev commented on app/assets/stylesheets/common.scss:
```diff
@@ -1003,3 +1003,23 @@ img.trace_image {
+
+.heatmap-labels {
+position: absolute;
+  top: 2px;
+  left: 17px;
+  padding-top: 21px;
```
Why do you need both `top` and `padding-top`?

kcne replied:
I had issues while making this work on UI with positioning. Fixed now, removed padding top and relied on top property completely instead. Everything works properly. Marking as resolved.

AntonKhorev commented on app/assets/stylesheets/common.scss:
```diff
@@ -1003,3 +1003,23 @@ img.trace_image {
+
+.heatmap {
+  margin-left: 50px;
+  height: 100%;
+  z-index: 0;
```
Do these two properties do anything?

kcne replied:
I thought they did as precaution but they don't. Removed now and everything works ok. Thank you.

AntonKhorev commented on app/views/users/show.html.erb:
```diff
@@ -1,3 +1,7 @@
+<% content_for :head do %>
+  <%= stylesheet_link_tag "heatmap" %>
+  <%= javascript_include_tag "heatmap" %>
```
The linked files don't exist yet. And the stylesheet won't exist later too.

kcne replied:
Removed from this commit and added later when the files are added. Thank you.

AntonKhorev commented on app/views/users/show.html.erb:
```diff
@@ -233,6 +237,12 @@
+  <div class="overflow-auto" style="max-height: 500px;">
+    <%= tag.div(:id => "cal-heatmap", :data => { :heatmap => @heatmap_data.to_json }) %>
+  </div>
+<% end %>
```
You overwrite this later almost entirely. I'd keep `if` because we will not output the heatmap [under some conditions](https://github.com/openstreetmap/openstreetmap-website/pull/5402#issuecomment-2602166082).

`style="max-height: 500px;"` shouldn't be here.

kcne replied:
Okay, kept the if statement as it is now and also removed styling mentioned and `console.warn` statment from the javascript code. Thank you.

kcne replied:
When we decide the conditions for renedering we can refactor the if statement as necessary. Thank you.

AntonKhorev commented on app/views/users/show.html.erb:
```diff
@@ -237,11 +237,29 @@
-<% if @heatmap_data.present? %>
-  <div class="overflow-auto" style="max-height: 500px;">
-    <%= tag.div(:id => "cal-heatmap", :data => { :heatmap => @heatmap_data.to_json }) %>
+<div class="small">
```
What does `class="small"` do?

kcne replied:
Removed that div entirely, thank you. I thought  it would help make it responsive on small screens in bootstrap grid, but apparently it doesn't do anything in this case.

AntonKhorev commented on app/views/users/show.html.erb:
```diff
@@ -237,11 +237,29 @@
-  <div class="overflow-auto" style="max-height: 500px;">
-    <%= tag.div(:id => "cal-heatmap", :data => { :heatmap => @heatmap_data.to_json }) %>
+<div class="small">
+  <div class="row">
+    <div class="col overflow-auto">
+      <div id="cal-heatmap-wrapper" class="heatmap-wrapper">
```
Why wouldn't you use a single flex container that puts weekday labels next to the heatmap? No position absolute/relative, probably only one pixel value of top padding for the labels to align them.

kcne replied:
Refactored the code to use flexbox instead. It is indeed much simpler, as I could get rid of some custom css classes. Thanks once again.

AntonKhorev commented on test/controllers/users_controller_test.rb:
```diff
@@ -333,4 +333,97 @@ def test_auth_failure_callback
+  end
+
+  def test_heatmap_rendering
+    # Test user with no changesets
+    user_without_changesets = create(:user)
+    get user_path(user_without_changesets)
+    assert_response :success
+    assert_select "div#cal-heatmap", 1
```
This fails because `@heatmap_data.present?` is false when `@heatmap_data` is an empty array.

kcne replied:
Updated the test to reflect latest changes. Thank you.

AntonKhorev commented on app/views/users/show.html.erb:
```diff
@@ -233,6 +237,30 @@
+        <!-- Heatmap -->
+        <div id="cal-heatmap" class="heatmap"
+          data-heatmap="<%= @heatmap_data.to_json %>"
+          data-site_color_scheme="<%= preferred_color_scheme(:site) %>">
```
The color scheme is already present in `data-bs-theme` on `<html>` ("auto" if that attribute is missing).

AntonKhorev commented on config/locales/en.yml:
```diff
@@ -3300,6 +3300,10 @@ en:
       centre_map: Centre map here
+    heatmap:
+      tooltip:
+        contributions: "%{count} contributions on %{date}"
```
You need `one` and `other` versions with `%{count}`.

kcne replied:
Updated keys and js code respectively. Now we got `no_contributions`, `contributions` with `%{count}` and `one_contribution`.

AntonKhorev commented on config/locales/en.yml:
```diff
@@ -3302,6 +3302,11 @@ en:
       centre_map: Centre map here
+    heatmap:
+      tooltip:
+        one_contribution: "1 contribution on %{date}"
+        contributions: "%{count} contributions on %{date}"
```
```suggestion        
        contributions:
          one: "%{count} contribution on %{date}"
          other: "%{count} contributions on %{date}"
```

kcne replied:
Resolved.

AntonKhorev commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,99 @@
+  const localizedDate = I18n.l("date.formats.long", date);
+
+  if (value > 0) {
+    return value === 1 ?
+      I18n.t("javascripts.heatmap.tooltip.one_contribution", { date: localizedDate }) :
+      I18n.t("javascripts.heatmap.tooltip.contributions", { count: value, date: localizedDate });
```
```suggestion
    return I18n.t("javascripts.heatmap.tooltip.contributions", { count: value, date: localizedDate });
```

kcne replied:
Resolved.

kcne commented:
The test failing here doesn't have nothing to do with the changes introduced in this PR. I tried force pushing again to rerun checks but again got failing test on different version of Ruby this time, which is weird. Is there some kind of bug in CI/CD right now or should I push again to rerun checks?

AntonKhorev commented on yarn.lock:
```diff
@@ -4,24 +4,24 @@
 "@aashutoshrathi/word-wrap@^1.2.3":
   version "1.2.6"
-  resolved "https://registry.yarnpkg.com/@aashutoshrathi/word-wrap/-/word-wrap-1.2.6.tgz#bd9154aec9983f77b3a034ecaa015c2e4201f6cf"
+  resolved "https://registry.npmjs.org/@aashutoshrathi/word-wrap/-/word-wrap-1.2.6.tgz"
```
Now there's a ton of `registry.yarnpkg.com` -> `registry.npmjs.org` changes.

kcne replied:
Must have messed this up during conflict. I will try to restore the file and do rebase again. Thank you.

AntonKhorev replied:
Now there are no updates to `yarn.lock`.

kcne replied:
Should be resolved now.

kcne commented:
I've resolved the conflicts in `yarn.lock` again. Since changes keep getting merged into master, I keep running into conflicts. Are there any remaining issues to address, or is this good to merge now?

tomhughes commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,100 @@
+//= require d3/dist/d3
+//= require cal-heatmap/dist/cal-heatmap
+//= require popper
```
I don't understand how this works? Shouldn't it be `@popperjs/core` or something here? How is it finding `popper` and is it actually finding the version we think it is?

kcne replied:
This import is the same as the one inside `application.js` and was not introduced in this PR. It's provided by the [popper_js gem](https://github.com/glebm/popper_js-rubygem), which looking at the repo automatically supplies popper.js assets to rails.

AntonKhorev commented:
With #5654 it's possible to translate the tooltips with language-specific pluralizations:

```
    heatmap:
      tooltip:
        contributions:
          one: "%{count} правка %{date}"
          few: "%{count} правки %{date}"
          many: "%{count} правок %{date}"
          other: ""
        no_contributions: "Нет правок %{date}"
```

![image](https://github.com/user-attachments/assets/046d0758-acc7-4f18-8022-9d71ef19f07a) ![image](https://github.com/user-attachments/assets/066270ae-458a-460b-a9e7-31267e09f88c) ![image](https://github.com/user-attachments/assets/b665eeca-0b49-4843-82cb-fedded5ce871)

Otherwise only `one` for `count==1` and `other` rules are supported.

Do we merge #5654 first?

kcne commented:
Now that #5654 was merged, I think current implementation is already setup to support this. Please let me know if i need to add anything else to support it. Thank you.

tomhughes commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,100 @@
+        type: "month",
+        gutter: 4,
+        label: {
+          text: (timestamp) => I18n.strftime(new Date(2000, new Date(timestamp).getMonth(), 15), "%b"),
```
If we've constructed a date from the timestamp why do you then extract the month from that and build a new date? If the format only uses the month the rest of the date doesn't matter?

It might a bit faster to just lookup `I18n.t("date.abbr_month_names")` once before the paint call and then index that array with the month number rather than calling strftime repeatedly - just beware of 0/1 based indexing!

kcne replied:
You're right. Updated the code. Thank you.

tomhughes commented on app/assets/javascripts/heatmap.js:
```diff
@@ -0,0 +1,102 @@
+//= link cal-heatmap/dist/cal-heatmap.css
```
I think the best solution is probably to create `app/assets/heatmap.scss` which can require that, and also include the heap rule you added to `common.scss` and then have the view load `heatmap` rather than `cal-heatmap/dist/cal-heatmap`.

kcne replied:
Thanks for the feedback, Tom. Since the code inside `common.scss` was unnecessary, I removed it completely, linked the `.css` file to `manifest.js`, and included it in the view directly. Let me know if any further modifications are needed.

tomhughes reviewed (approved):
I might still have added a local stylesheet with a require just so there was consistency between the javascript and the styles but it doesn't really matter so I think this is good to go now - thanks for your work on it!

verhovsky commented:
You need to hide days that haven't happened, like GitHub. Every time you open it, it looks like the user stopped contributing a few weeks ago.

HolgerJeromin commented:
> You need to hide days that haven't happened, like GitHub. Every time you open it, it looks like the user stopped contributing a few weeks ago.

Would be nice. 
But I also did not found a way to do that with the used library...
